---
title: "Lucien 16S rRNA data"
author: "Jadi"
date: "22/02/24"
output: html_document
---
```{r}
knitr::opts_chunk$set(include = F)
knitr::opts_chunk$set(eval = T)
knitr::opts_chunk$set(echo = F)
knitr::opts_chunk$set(message = F)
knitr::opts_chunk$set(warning = T)
#knitr::opts_chunk$set(fig.cap = "...")
```

```{r}
library(vegan)
library(tidyverse)
library(RColorBrewer)
library(lme4)
library(multcomp)
library(lsmeans)
library(mvabund) 
library(ellipse)
library(gplots)
library(pheatmap)
library(ggsci)
library(ggpubr)
library(gtable)
library(sandwich)
library(scales)
library(DESeq2)
library(gtools)
#library(QsRutils)
#library(iNEXT)
library(stringr)
library(phyloseq)


```

Removed manually the Confidence Values from original table and Re labelled the samples names!!!!
```{r}
ASV_table_raw <- read.csv("outputs/files/Split_AllSamples_unoise_otu_table_BLCA_filtered2.csv")
head(ASV_table_raw)
ASV_table_raw[1:5,1:5]
dim(ASV_table_raw) #139  38

unique(ASV_table_raw$ASV_ID) # 
```

Create data table - no taxa

```{r}

# OTUs in columns, samples in rows
ASV_table_data <- ASV_table_raw %>%
  dplyr::select(ASV_ID:S_4_9) %>%
  column_to_rownames(var = "ASV_ID") %>%
  t() %>%
  as.data.frame()

ASV_table_data[1:10, 1:10]
dim(ASV_table_data) # 30 139
sum(rowSums(ASV_table_data[2:30,])) #1193166
rownames(ASV_table_data)

```

Metadata

Need to change the order of the metadata table 

```{r}
fact <- read.csv("outputs/files/20240105_asparagopsis_16S_metadata.csv", row.names = 1)
fact$Label_Sequencing <- rownames(fact)
ASV_table_data <- ASV_table_data[order(match(rownames(ASV_table_data), fact$Label_Sequencing)),]
ASV_table_data[1:10, 1:10]

rownames(ASV_table_data) == fact$Label_Sequencing

```

Check raw dataset for zeros across all samples in both datasets 
```{r}
# zeros generated by removing NSW samples from the dataset
dim(ASV_table_data[, colSums(ASV_table_data) == 0]) # 0 UNI with zero counts
# remove these UNI sequences
ASV_table_data <- ASV_table_data[, colSums(ASV_table_data) > 0] 
dim(ASV_table_data) #  30 139 unique sequences after removal of UNI with zero counts
# check if ASV table and fact have samples in the same order

rownames(fact) <- rownames(ASV_table_data)
rownames(fact) == rownames(ASV_table_data)
fact$Label_Sequencing == rownames(ASV_table_data)
```

Create taxonomy table for whole dataset - no filtering
The taxonomy object with split lineages and "Unassigned" for empty spaces.
```{r}
dim(ASV_table_data) # 30 139 the taxonomy was manually split
ASV_taxa_ALL <- read.csv("outputs/files/Taxa_all.csv")

#Create the taxonomy object with  "Unassigned" for empty spaces.
ASV_taxa <- ASV_taxa_ALL
head(ASV_taxa)

ASV_taxa$Domain <- as.character(ASV_taxa$Domain)
ASV_taxa$Class <- as.character(ASV_taxa$Class)
ASV_taxa$Phylum <- as.character(ASV_taxa$Phylum)
ASV_taxa$Order <- as.character(ASV_taxa$Order)
ASV_taxa$Family <- as.character(ASV_taxa$Family)
ASV_taxa$Genus <- as.character(ASV_taxa$Genus)
ASV_taxa$Species <- as.character(ASV_taxa$Species)

ASV_taxa <- mutate_all(ASV_taxa, list(~na_if(.,"")))
ASV_taxa
ASV_taxa[is.na(ASV_taxa)] <- "Unassigned"
ASV_taxa

ASV_taxa$Domain <- as.factor(ASV_taxa$Domain)
ASV_taxa$Class <- as.factor(ASV_taxa$Class)
ASV_taxa$Phylum <- as.factor(ASV_taxa$Phylum)
ASV_taxa$Order <- as.factor(ASV_taxa$Order)
ASV_taxa$Family <- as.factor(ASV_taxa$Family)
ASV_taxa$Genus <- as.factor(ASV_taxa$Genus)
ASV_taxa$Species <- as.factor(ASV_taxa$Species)

#ASV_table_data <- ASV_table_data %>% rownames_to_column("ASV_ID") 
tASV_table_data <- t(ASV_table_data) %>% as.data.frame() %>%  rownames_to_column("ASV_ID") 
ASV_taxa_noZeroes <- ASV_taxa %>%
  filter(ASV_ID %in% tASV_table_data$ASV_ID) %>%
  dplyr::select(ASV_ID, everything()) %>%
  column_to_rownames(var = "ASV_ID")
dim(ASV_taxa_noZeroes) # 139    7
head(ASV_taxa_noZeroes)

str(ASV_taxa_noZeroes)


unique(ASV_taxa_noZeroes$Domain) 
unique(ASV_taxa_noZeroes$Phylum) 
unique(ASV_taxa_noZeroes$Class) 
unique(ASV_taxa_noZeroes$Family) 
unique(ASV_taxa_noZeroes$Order) 
unique(ASV_taxa_noZeroes$Genus) 
unique(ASV_taxa_noZeroes$Species) 

#write.csv(ASV_taxa, "outputs/files/taxa_data.csv")

```

How many ASV are unassigned or Archea from the Domain level:

```{r}
data_NoUnassigned <- ASV_taxa_noZeroes %>% 
  filter(!Domain == "Unclassified")

dim(data_NoUnassigned) # 

data_Unassigned <- ASV_taxa_noZeroes %>% 
  filter(Domain == "Unclassified")
dim(data_Unassigned) # 2  ASV are Unassigned 


data_Unassigned_Archaea <- ASV_taxa_noZeroes %>% 
  filter(!Domain == "Bacteria")
dim(data_Unassigned_Archaea) # 2 ASV are Unassigned or Archaea

data_NoArchaea <- ASV_taxa_noZeroes %>% 
  filter(!Domain == "Archaea")

dim(data_NoArchaea) # 

data_Archaea <- ASV_taxa_noZeroes %>% 
  filter(Domain == "Archaea")
dim(data_Archaea) # 0 ASV are Archea

```

Basic stats on sequencing depth
```{r}
# total number of reads per sample
rowSums(ASV_table_data)
summary(rowSums(ASV_table_data))
  #   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  # 25248   36093   41034   40957   46390   56843 

# range of differences in sequencing depth
max((rowSums(ASV_table_data))) / min(rowSums(ASV_table_data))
# 2.25

```

```{r}

ASV_taxa_noZeroes$ASV_ID <- rownames(ASV_taxa_noZeroes)
ASV_table_data_tax <- left_join(tASV_table_data, ASV_taxa_noZeroes, by = "ASV_ID")

#### remove  Unclassified #####

data_table_noUnclass <- ASV_table_data_tax %>% 
  filter(!Domain == "Unclassified")
dim(ASV_table_data_tax) # 139 38
dim(data_table_noUnclass) # 137 38
str(data_table_noUnclass)

write.csv(data_table_noUnclass, "outputs/files/Raw_noUnassigned_tax.csv")

```



Define colors for rarefaction curves
```{r}
library(randomcoloR)
n <- 11
myPalette_treatment <- distinctColorPalette(n)

n2 <- 92
myPalette_samples <- distinctColorPalette(n2)

```

Rarefaction curves
```{r}
# Calculate minimum read count achieved across samples
raremax <- min(rowSums(ASV_table_data)) # smaller sample S_4_7: 25248
head(ASV_table_data)
tail(ASV_table_data)

#ASV_table_data$sample_name <- rownames(ASV_table_data)

ASV_table_fact <- ASV_table_data %>%
  t() %>% as.data.frame() %>% rownames_to_column("ASV_ID") %>% 
  gather(key = Label_Sequencing, value = count, S_0_0:S_4_31) %>%
  left_join(., fact, by = "Label_Sequencing")
head(ASV_table_fact)
str(ASV_table_fact) # 


str(ASV_table_fact$count)


# Plot for number of sequeces per sample 
plot_numbSeq <-  ASV_table_fact %>%
  group_by(Label_Sequencing, Treatment) %>%
  summarise(sum_seqs = sum(count, na.rm = TRUE)/1000) %>%
  ggplot(aes(x = Label_Sequencing, y = sum_seqs)) +
  geom_bar(aes(reorder(Label_Sequencing, -sum_seqs), fill = Treatment), stat = "identity", color = "black", size = 0.05) +
  #facet_wrap(~ TOOKSEAWEED, scales = "free", ncol = 2) +
  scale_fill_manual(values = myPalette_treatment) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) 
plot_numbSeq
#ggsave(plot = plot_numbSeq, filename = "outputs/figures/numberSamples.pdf", width = 8, height = 5)

```

CONSIDER REMOVIING Phs_PC_D4_04 IN THE SUMMER RUN GROUP!!!! THE SEQUENING IS PRETTY BAD FOR THIS SAMPLE


Define colors for rarefaction curves
```{r}
#EXTRA
#PRE
#POST

myPalette_stage <- c("skyblue1", "violetred", "orange")
```
Rarefaction curve
```{r}
# pdf("outputs/figures/rarefaction_curve.pdf")
# raref.curve <- rarecurve(ASV_table_data, ylab = "ASV Count") # Long step
# dev.off()

```
REMOVE sample XXX --> use this line if there are samples that need to be removed.
```{r}
dim(ASV_table_data) #

ASV_table_data_2 <- ASV_table_data[!row.names(ASV_table_data) %in% "S_0_0",]
dim(ASV_table_data_2)
fact2 <- fact[!row.names(fact) %in% "S_0_0",]
dim(fact2)

```

Data normalisation using DESeq2:
counts divided by sample-specific size factors determined by median ratio of gene counts relative to geometric mean per gene

DISEASES samples
```{r}
library(DESeq2)

# # two-way analysis with 2 levels each (treatment + time)
# # countData = ASV counts with samples in columns  
# # colData = Sample meta data (factors, etc)  
# # design = the right hand side of a GLM formula
# # 
ASV_deseq <- DESeqDataSetFromMatrix(countData = t(ASV_table_data_2),
                                           colData = fact2,
                                           design = ~ Treatment)

# positive counts normalisaton (accounts for zero-inflation)
ASV_deseq <- estimateSizeFactors(ASV_deseq, type = "poscounts")
sizeFactors(ASV_deseq)

# size-factor corrected data are calculated by dividing the raw counts by the sample size factor and adding 0.5 to correct for the zeros
ASV_deseq_norm <- sapply(row.names(ASV_deseq), function(x){
  plotCounts(ASV_deseq, x, "Treatment", returnData = TRUE, normalized = TRUE)$count

})
rownames(ASV_deseq_norm) <- colnames(ASV_deseq)

# remove the addition of 0.5 to all entries
ASV_deseq_norm <- ASV_deseq_norm - 0.5
ASV_deseq_norm[1:10, 1:10]

# make dataset with integers
ASV_deseq_norm <- round(ASV_deseq_norm)
ASV_deseq_norm[1:10, 1:10]

# check for zeros across all samples (produced by normalisation)
dim(ASV_deseq_norm[, colSums(ASV_deseq_norm) == 0]) # 3

# check singletons
dim(ASV_deseq_norm[, colSums(ASV_deseq_norm) == 1]) # 5 singletons - keep these

#save normalised table
#write.csv(ASV_deseq_norm, "outputs/files/normalised_ASV_table.csv")
#write.csv(ASV_deseq_norm, "outputs/files/normalised_ASV_table_noNC.csv")
```

Good's coverage index

```{r}
library(QsRutils)

dat_norm <-  read.csv("outputs/files/normalised_ASV_table_noNC.csv", row.names = 1)
head(dat_norm)
# # samples as rows and OTUs as columns
goods_result <- goods(dat_norm)
summary(goods_result$goods)
 # Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 #  99.96   99.97   99.98   99.98   99.99  100.00 
 
mean(goods_result$goods) # 99.98
round(sd(goods_result$goods), 2) # 0.01


```

Multivariate analyses 
```{r}
#### NMDS ####
library(vegan)
library(RColorBrewer)
library(MASS)
library(randomcoloR)

#ASV_deseq_norm <- read.csv("outputs/files/normalised_ASV_table.csv", row.names = 1)

ASV_deseq_norm <- read.csv("outputs/files/normalised_ASV_table_noNC.csv", row.names = 1)
ASV_deseq_norm[1:10, 1:10]
sqrt_ASV_deseq_norm <- sqrt(ASV_deseq_norm)
# Calculate Bray-Curtis dissimilarity matrix
dat_bc <- vegdist(sqrt_ASV_deseq_norm, method = "bray")
dat_mds <- metaMDS(dat_bc, trymax = 1000)
dat_mds # stress = 

# NMDS plot
MDS_xy <- data.frame(dat_mds$points)
MDS_xy$Treatment <- fact2$Treatment

# NMDS plot with Treatment and Time as individual factors
MDS_xy_Treat <- data.frame(MDS1 = dat_mds$point[,1], MDS2 = dat_mds$point[,2], Treatment = as.factor(fact2$Treatment))
unique(fact2$Treatment)
library(viridis)
n <- 5
myPalette_treatment <- distinctColorPalette(n)

# plot and save
colour_values <- myPalette_treatment
colour_values_treatment <- c("#66E79E",  "#E2A754" ,"#D965B2" ,"#DAE25F" ,"#BF47E2", "#DC5F5B")

myPalette <- c("#C0ECB3", "#7A8985", "#D0DFE0", "#66E79E", "#DED1A0", "#71AA64" ,"#82E754", "#78E1D8" ,"#7D73D4", "#D4A6DC", "#DD9FA7", "#DC5F5B", "#83B6E1", "#E2A754" ,"#D965B2" ,"#DAE25F" ,"#BF47E2")

plot <- ggplot(MDS_xy_Treat, aes(MDS1, MDS2)) +
  geom_point(aes(colour = Treatment),
                 size = 4) +
 # scale_color_viridis(discrete = TRUE, option = "D")+
  #scale_fill_viridis(discrete = TRUE) +
  scale_color_manual(values = colour_values_treatment) +
  xlab("NMDS1") +
  ylab("NMDS2") +
  theme_bw() +
  theme(panel.grid = element_blank()) +
  theme(aspect.ratio = 1) 
plot
#ggsave(plot = plot, filename = "outputs/figures/NMDS_allOTUs_noNC.pdf", width = 7, height = 7)
#ggsave(plot = plot, filename = "outputs/figures/NMDS_allOTUs_wNC.pdf", width = 7, height = 7)
```


Calculate the Relative Abundances from the count table with removed "Unclassified" at the Domain level. KEPT the Archea
```{r}
data_table_noUnclass <- read.csv("outputs/files/Raw_noUnassigned_tax.csv", row.names = 1)
data_table_noUnclass_counts <- data_table_noUnclass[,1:31] %>%  as.data.frame()
data_table_noUnclass_taxa <- data_table_noUnclass[,32:38] %>%  as.data.frame()

tdata_table_noUnclass_counts <- data_table_noUnclass_counts %>%  t() %>% as.data.frame() 
colnames(tdata_table_noUnclass_counts) <- tdata_table_noUnclass_counts[1,] %>% as.data.frame() 
dim(tdata_table_noUnclass_counts) #87 2630
tdata_table_noUnclass_counts <- tdata_table_noUnclass_counts[2:31,] %>% as.data.frame() 
str(tdata_table_noUnclass_counts)
write.csv(tdata_table_noUnclass_counts, "outputs/files/Transposed_Rawdata_noUnass.csv")
tdata_table_noUnclass_counts <- read.csv("outputs/files/Transposed_Rawdata_noUnass.csv", row.names = 1)

data_rel <- tdata_table_noUnclass_counts/rowSums(tdata_table_noUnclass_counts)
head(data_rel)
rowSums(data_rel) 

write.csv(data_rel, "outputs/files/RelativeAbundances.csv")

```


Taxa plot

```{r}

#Merge ASV table and taxa table 
data_RelAbund <- data_rel %>% 
  t() %>% 
   as.data.frame()
head(data_RelAbund)
write.csv(data_RelAbund, "outputs/files/ASV_table_RelativeAbund_Transposed.csv")
data_RelAbund <-  read.csv("outputs/files/ASV_table_RelativeAbund_Transposed.csv",row.names = 1 )
data_RelAbund <- data_RelAbund %>%  rownames_to_column("ASV_ID")
ASV_taxa <- data_table_noUnclass_taxa %>%  rownames_to_column(var = "ASV_ID")

head(data_RelAbund)

head(ASV_taxa)

data_RelAbund$ASV_ID
ASV_taxa$ASV_ID

rownames(ASV_taxa) <- data_RelAbund$ASV_ID
rownames(data_RelAbund) <- data_RelAbund$ASV_ID

rownames(ASV_taxa) == rownames(data_RelAbund)
ASV_taxa$ASV_ID <- rownames(ASV_taxa)

data_RelAbund_tax <- data_RelAbund %>% 
  left_join(., ASV_taxa, by = "ASV_ID")
head(data_RelAbund_tax)

dim(data_RelAbund_tax) #  137 38
#### check for mitochondria and chloroplast sequences ####
#### Inspect data for entries that needs to be removed (e.g. Archaea, Chloroplasts, Mitochondria) ####
sort(unique(data_RelAbund_tax$Domain)) #      "Bacteria"    


```

Set colours for plot 
```{r}
library(randomcoloR)
n <- 90
palette <- distinctColorPalette(n)
# n <- 70
# qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
# col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))
# pie(rep(1,n), col=sample(col_vector, n))
# color palette
taxaPalette <- distinctColorPalette(n)
getPalette_taxa <- colorRampPalette(taxaPalette)
```

Data long 

```{r}
dat_long <- data_RelAbund %>%
  #rownames_to_column(var = "ASV_ID") %>%
  tidyr::gather(key = "Label_Sequencing", value = "RelativeAbundance", S_3_2:S_4_31) %>%
  mutate(ASV_sample = paste(.$ASV_ID, .$Label_Sequencing, sep = "_"))
head(dat_long)

# merge tables
dat_final <- dat_long %>%
  dplyr::select(ASV_ID, Label_Sequencing, RelativeAbundance, ASV_sample) %>%
  left_join(., fact, by = "Label_Sequencing") %>%
  left_join(., data_RelAbund_tax, by = "ASV_ID")
head(dat_final)
tail(dat_final)


```


Plot TAXA 

Genus - replicates
 
```{r}
#Genus
# calculate sum of absolute abundance of each Genus per replicate within a treatment
Genus_NoUnassigned <- data.frame(dat_final %>%
                                  dplyr::group_by(Label_Sequencing, Treatment, Genus) %>%
                                  dplyr::summarise(totalRelAbund = sum(RelativeAbundance)))

unique(Genus_NoUnassigned$Genus) # 82

counts_Genus_Top <- Genus_NoUnassigned %>%
  mutate(Genus_updated = ifelse(totalRelAbund < 0.05, "Low abundant Genus < 0.05", as.character(Genus_NoUnassigned$Genus)))
unique(counts_Genus_Top$Genus_updated) # 11

taxaPalette <- getPalette_taxa(14)

plotUnassGenus <- ggplot(counts_Genus_Top, aes(Label_Sequencing, totalRelAbund)) +
  geom_bar(stat="identity",aes(fill = Genus_updated)) +
  theme_bw() +
  theme(axis.text.x = element_text(size = 10)) +
  theme(plot.title = element_text(hjust = 0.5)) +
  theme(legend.position = "bottom") + theme(legend.title = element_blank()) + 
  theme(legend.title = element_text(size = 12)) +
  #scale_y_continuous(expand = c(0,0), labels = scales::percent) +
  scale_fill_manual("legend", values = taxaPalette) +
  xlab("sample") +
  ylab("Relative Abundance") + 
  facet_wrap(.~ Treatment, scales = "free", ncol = 2) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  theme(panel.grid = element_blank())
plotUnassGenus
ggsave(plot = plotUnassGenus , filename = "outputs/figures/Genus_treatment_0.05.pdf", width = 7, height = 10)
```

Genus - average by treatment
 
```{r}
#Genus
# calculate sum of absolute abundance of each Genus per replicate within a treatment
Genus_NoUnassigned_AVG <- data.frame(Genus_NoUnassigned %>%
                                  dplyr::group_by(Treatment, Genus) %>%
                                  dplyr::summarise(AvgRelAbund = mean(totalRelAbund)))

unique(Genus_NoUnassigned_AVG$Genus) # 1000

counts_Genus_Top_AVG <- Genus_NoUnassigned_AVG %>%
  mutate(Genus_updated = ifelse(AvgRelAbund < 0.05, "Low abundant Genus < 0.05", as.character(Genus_NoUnassigned_AVG$Genus)))
unique(counts_Genus_Top_AVG$Genus_updated) # 7

taxaPalette <- getPalette_taxa(9)

plotUnassGenus_AVG <- ggplot(counts_Genus_Top_AVG, aes(Treatment, AvgRelAbund)) +
  geom_bar(stat="identity",aes(fill = Genus_updated)) +
  theme_bw() +
  theme(axis.text.x = element_text(size = 10)) +
  theme(plot.title = element_text(hjust = 0.5)) +
  theme(legend.position = "bottom") + theme(legend.title = element_blank()) + 
  theme(legend.title = element_text(size = 12)) +
  #scale_y_continuous(expand = c(0,0), labels = scales::percent) +
  scale_fill_manual("legend", values = taxaPalette) +
  xlab("treatment") +
  ylab("Relative Abundance") + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  theme(panel.grid = element_blank())
plotUnassGenus_AVG
ggsave(plot = plotUnassGenus_AVG , filename = "outputs/figures/Genus_treatment_time_Top_AVERAGE.pdf", width = 7, height = 5)
```

Heatmaps:
All samples data fpr heatmap by Family - top 30
```{r}
# library(pheatmap)
# 
# # top abundance
# topAbundance <- dat_final %>%
#   dplyr::group_by(Treatment, ASV_ID) %>%
#   dplyr::summarise(mean_ab = mean(RelativeAbundance)) %>%
#   dplyr::ungroup() %>%
#   dplyr::group_by(ASV_ID) %>%
#   dplyr::summarise(sum_ASV = sum(mean_ab)) %>%
#   arrange(desc(sum_ASV)) %>%
#   dplyr::top_n(n = 50, wt = sum_ASV)
# head(topAbundance)
# 
# Genus_top30 <- dat_final %>%
#   mutate(ASV_Genus = paste(ASV_ID, Genus, sep = "__")) %>%
#     filter(ASV_ID %in% topAbundance$ASV_ID) %>%
#   dplyr::select(ASV_Genus, Label_Sequencing, Treatment, RelativeAbundance) %>%
#   dplyr::group_by(treatment, ASV_Genus) %>%
#   dplyr::summarise(mean_group = mean(RelativeAbundance)) %>%
#   spread(key = treatment, value = mean_group) %>%
#   as.data.frame() %>%
#   column_to_rownames(var = "ASV_Genus")
# head(Genus_top30)

```

```{r}
 
# Quantile breaks for heatmap
# most of the data comprise low abundant sequences
# Quantile breaks for heatmap
# quantile_breaks <- function(xs, n = 10) {
#   breaks <- quantile(xs, probs = seq(0, 1, length.out = n))
#   breaks[!duplicated(breaks)]
# }
# 
# mat_breaks_avg <- quantile_breaks(as.matrix(Genus_top30), n = 1000)
#    
# 
# pheatmap(
#   mat               = as.matrix(Genus_top30),
#   color             = colorRampPalette(rev(brewer.pal(n = 9, name =  "Spectral")))(length(mat_breaks_avg) - 1),
#   #color             = hcl.colors(700, "BluYl"),
#   #color             = viridis(length(mat_breaks_avg) - 1),
#   breaks            = mat_breaks_avg,
#   clustering_method = 'average',
#   cluster_cols = T,
#   cluster_rows = T,
#   drop_levels = TRUE,
#   border_color = NA,
#   fontsize = 8,
#   display_numbers = F,
#   filename = "outputs/figures/heatmap_top30.pdf", width = 15, height = 10
#   )
# 
# 
# Genus_top30_AVG <- dat_final %>%
#   mutate(ASV_Genus = paste(ASV_ID, Genus, sep = "__")) %>%
#     filter(ASV_ID %in% topAbundance$ASV_ID) %>%
#   dplyr::select(ASV_Genus, Label_Sequencing, Treatment, RelativeAbundance) %>%
#   dplyr::group_by(Treatment, ASV_Genus) %>%
#   dplyr::summarise(mean_group = mean(RelativeAbundance)) %>%
#   spread(key = Treatment, value = mean_group) %>%
#   as.data.frame() %>%
#   column_to_rownames(var = "ASV_Genus")
# head(Genus_top30_AVG)
#  
# # Quantile breaks for heatmap
# # most of the data comprise low abundant sequences
# # Quantile breaks for heatmap
# quantile_breaks <- function(xs, n = 10) {
#   breaks <- quantile(xs, probs = seq(0, 1, length.out = n))
#   breaks[!duplicated(breaks)]
# }
# 
# mat_breaks_avg <- quantile_breaks(as.matrix(Genus_top30_AVG), n = 1000)
#    
# 
# pheatmap(
#   mat               = as.matrix(Genus_top30_AVG),
#   color             = colorRampPalette(rev(brewer.pal(n = 9, name =   "Spectral")))(length(mat_breaks_avg) - 1),
#   #color             = hcl.colors(700, "BluYl"),
#   #color             = viridis(length(mat_breaks_avg) - 1),
#   breaks            = mat_breaks_avg,
#   clustering_method = 'average',
#   cluster_cols = T,
#   cluster_rows = T,
#   drop_levels = TRUE,
#   border_color = NA,
#   fontsize = 8,
#   display_numbers = T,
#   filename = "outputs/figures/heatmap_top30_AVG.pdf", width = 5, height = 10
#   )

```


1) blast PROBS vs the ASV
results from blast:


************************************************************
************************************************************
************************************************************
************************************************************
************************************************************
                PROBIOTICS - ASV PLOTS
************************************************************
************************************************************
************************************************************
************************************************************
************************************************************

Heatmap with probiotic ASVs samples



```{r}

# ASV probiotics
ASV_probs <- dat_final %>%
  filter(ASV_ID %in% c("Zotu14","Zotu25", "Zotu37"))
  
head(ASV_probs)

Genus_top50_probASV <- dat_final %>%
  mutate(ASV_Genus = paste(ASV_ID, Genus, sep = "__")) %>%
  filter(ASV_ID %in% ASV_probs$ASV_ID) %>%
  #filter(antibiotic %in% c("AB")) %>% 
  dplyr::select(ASV_Genus, Label_Sequencing, RelativeAbundance) %>%
  dplyr::group_by(Label_Sequencing, ASV_Genus) %>%
  dplyr::summarise(mean_group = mean(RelativeAbundance)) %>%
  spread(key = Label_Sequencing, value = mean_group) %>%
  as.data.frame() %>%
  column_to_rownames(var = "ASV_Genus")
head(Genus_top50_probASV)

Genus_top50_probASV_AVG <- dat_final %>%
  mutate(ASV_Genus = paste(ASV_ID, Genus, sep = "__")) %>%
  filter(ASV_ID %in% ASV_probs$ASV_ID) %>%
  #filter(antibiotic %in% c("AB")) %>% 
  dplyr::select(ASV_Genus, Treatment, RelativeAbundance) %>%
  dplyr::group_by(Treatment, ASV_Genus) %>%
  dplyr::summarise(mean_group = mean(RelativeAbundance)) %>%
  spread(key = Treatment, value = mean_group) %>%
  as.data.frame() %>%
  column_to_rownames(var = "ASV_Genus")
head(Genus_top50_probASV_AVG)

# Quantile breaks for heatmap
# most of the data comprise low abundant sequences
# Quantile breaks for heatmap
quantile_breaks <- function(xs, n = 10) {
  breaks <- quantile(xs, probs = seq(0, 1, length.out = n))
  breaks[!duplicated(breaks)]
}

mat_breaks_avg <- quantile_breaks(as.matrix(Genus_top50_probASV), n = 1000)
pheatmap(
  mat               = as.matrix(Genus_top50_probASV),
  color             = colorRampPalette(rev(brewer.pal(n = 9, name =  "PiYG")))(length(mat_breaks_avg) - 1),
 breaks            = mat_breaks_avg,
  clustering_method = 'average',
  cluster_cols = F,
  cluster_rows = T,
  drop_levels = TRUE,
  border_color = NA,
  fontsize = 8,
  display_numbers = F,
  filename = "outputs/figures/heatmap_probASV_noNC.pdf", width = 20, height = 4
  )

mat_breaks_avg <- quantile_breaks(as.matrix(Genus_top50_probASV_AVG), n = 1000)
pheatmap(
  mat               = as.matrix(Genus_top50_probASV_AVG),
  color             = colorRampPalette(rev(brewer.pal(n = 9, name =  "Spectral")))(length(mat_breaks_avg) - 1),
 breaks            = mat_breaks_avg,
  clustering_method = 'average',
  cluster_cols = F,
  cluster_rows = T,
  drop_levels = TRUE,
  border_color = NA,
  fontsize = 8,
  display_numbers = T, number_format = "%.1e",
  filename = "outputs/figures/heatmap_probASV_AVG_noNC.pdf", width = 19, height = 4
  )
```


Data LONG for ASV_probiotics


```{r}
Genus_sample <- data.frame(dat_final %>%
                              mutate(ASV_Genus = paste(ASV_ID, Genus, sep = "__")) %>%
                                  dplyr::group_by(Label_Sequencing, Treatment, ASV_Genus) %>%
                                  dplyr::summarise(totalRelAbund = sum(RelativeAbundance)))

Genus_AVG <- data.frame(Genus_sample %>%
                                  dplyr::group_by(Treatment, ASV_Genus) %>%
                                  dplyr::summarise(AvgRelAbund = mean(totalRelAbund)))

```


```{r}

data_final_probASV_ASV_Genus <- Genus_sample %>%
  #filter(treatment %in% c("probiotic_ThBc", "control_AB")) %>% 
  filter(ASV_Genus %in% c("Zotu14__Phaeobacter","Zotu37__Phaeobacter", "Zotu25__Pseudoalteromonas")) 


boxplot_probASV <- data_final_probASV_ASV_Genus %>%
  dplyr::group_by(Label_Sequencing, Treatment, ASV_Genus) %>%
  dplyr::summarise(mean_ab = mean(totalRelAbund)) %>%
  ungroup() %>%
  ggplot(aes(Treatment, mean_ab#, fill = treatment
             )) +
  geom_boxplot() +
  stat_summary(fun.y=mean, geom="point", shape=1, size=2, color="black", fill="black") +
  #scale_fill_viridis(discrete = TRUE, alpha=0.6, option="A") +
  geom_jitter(color="black", size=0.4, alpha=0.9) +
  theme_bw() +
  theme(axis.text.x = element_text(size = 8)) +
  theme(plot.title = element_text(hjust = 0.5)) +
  theme(legend.position = "bottom") +
  theme(legend.title = element_blank()) +
  scale_y_continuous(expand = c(0,0)) +
  xlab("Treatment") +
  ylab("Relative Abundance") + 
  facet_wrap(. ~ ASV_Genus, ncol = 3) +
  theme(panel.grid = element_blank()) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) 
 boxplot_probASV
ggsave(plot = boxplot_probASV, filename = "outputs/figures/boxplot_probASV_noNC.pdf", width =7, height = 4)

```
*******************************************************************
*******************************************************************
*******************************************************************
STATISTICAL ANALYSIS -- FROM HERE THE SCRIPT WILL HAVE TO BE MODIFIED BY YOU BASED ON YOUR DATA (COLUMNS/ROW NAMES IN YOUR TABLES, FACTOR NAMES, OBJECT NAMES YOU USED IN YOUR SCRIPT ETC...)

REMEBER TO RUN THE STATS WITH AND WITHOUT YOUR INOCULUM

*******************************************************************
*******************************************************************
*******************************************************************


PERMANOVA

```{r}

adonis2(NORMALISED_DATA ~ fact$Treatment)

```

MVABUND

```{r}
library(mvabund)
data_mvabund <- mvabund(ASV_table_data) #samples in rows and ASV is columns

dat_nb <- manyglm(data_mvabund ~ fact$treatment, family = "negative.binomial", data = fact, composition = TRUE) 

pdf("outputs/figures/meanvar_plot.pdf", width = 5, height = 5)
meanvar.plot(data_mvabund, legend = TRUE)
abline(0,1)
dev.off()

# if mean increases with the variance use a negative binomial distribution. Otherwise you could decide to use Poisson. I have never used poisson but it is really important that this assumption is met. If you want to check une factor at the time you can do as follows:
pdf("outputs/figures/ResvsFit_plot.pdf", width = 5, height = 5)
plot(dat_nb, n.vars = 100) #This will allow us to see if there is NO patterns, if there are patterns it means that we have a factor missing that can explain the experiment.
abline(0,1)
dev.off() # No obvious pattern among multiple plots - a good model fit.

pdf("outputs/figures/QQ_plot.pdf", width = 5, height = 5)
plot(dat_nb, which = 2) #we have to see haw the quantiles get organise in one line
dev.off()


#dat_aov <- anova(dat_nb, p.uni = "unadjusted")
#save.image("data.RData_unadj")
dat_aov$table #TREATMENT IS SIGNIFICANT
length(which(dat_aov$uni.p[2,] < 0.05)) #ASV with unadjusted pvalue


#dat_aov_adj <- anova(dat_nb, p.uni = "adjusted")
#save.image("data.RData_adj")
dat_aov_adj$table #TREATMENT IS SIGNIFICANT
length(which(dat_aov_adj$uni.p[2,] < 0.05)) #136
diff_05_treat_adj <- which(dat_aov_adj$uni.p[2,] < 0.05) 
names_diff_treat <- names(diff_05_treat_adj)

dat_diff_treat <- tASV_table_data[which(rownames(tASV_table_data) %in% names_diff_treat), ]
dim(dat_diff_treat) # 451 OTUs affected by treatment

#create a  new column named "response" and assign to each Zotu the factor they are affected by e.g. "treatment". Save the table.
dat_diff_treat$response <- "treatment"
#write.csv(dat_diff_treat, "outputs/files/rawData_diffabundantOTUs_mvabundGTDB_treatment_adj.csv")

```

Plot top abundant ASV from mvabund results 

```{r}
# top abundance
topAbundance <- dat_final %>%
  filter(ASV_ID %in% dat_diff_treat$ASV_ID) %>%
  dplyr::group_by(treatment, ASV_ID) %>%
  dplyr::summarise(mean_ab = mean(RelativeAbundance)) %>%
  dplyr::ungroup() %>%
  dplyr::group_by(ASV_ID) %>%
  dplyr::summarise(sum_ASV = sum(mean_ab)) %>%
  arrange(desc(sum_ASV)) %>%
  dplyr::top_n(n = 50, wt = sum_ASV)
head(topAbundance)


#dat_diff_treat$ASV_ID <- rownames(dat_diff_treat)

Genus_top50 <- dat_final %>%
  mutate(ASV_Genus = paste(ASV_ID, Genus, sep = "__")) %>%
  filter(ASV_ID %in% topAbundance$ASV_ID) %>%
  dplyr::select(ASV_Genus, sample_ID, RelativeAbundance) %>%
  dplyr::group_by(sample_ID, ASV_Genus) %>%
  dplyr::summarise(mean_group = mean(RelativeAbundance)) %>%
  spread(key = sample_ID, value = mean_group) %>%
  as.data.frame() %>%
  column_to_rownames(var = "ASV_Genus")
head(Genus_top50)

Genus_top50_AVG <- dat_final %>%
  mutate(ASV_Genus = paste(ASV_ID, Genus, sep = "__")) %>%
  filter(ASV_ID %in% topAbundance$ASV_ID) %>%
  dplyr::select(ASV_Genus, treatment, RelativeAbundance) %>%
  dplyr::group_by(treatment, ASV_Genus) %>%
  dplyr::summarise(mean_group = mean(RelativeAbundance)) %>%
  spread(key = treatment, value = mean_group) %>%
  as.data.frame() %>%
  column_to_rownames(var = "ASV_Genus")
head(Genus_top50_AVG)

# Quantile breaks for heatmap
# most of the data comprise low abundant sequences
# Quantile breaks for heatmap
quantile_breaks <- function(xs, n = 10) {
  breaks <- quantile(xs, probs = seq(0, 1, length.out = n))
  breaks[!duplicated(breaks)]
}

mat_breaks_avg <- quantile_breaks(as.matrix(Genus_top50), n = 1000)
pheatmap(
  mat               = as.matrix(Genus_top50),
  color             = colorRampPalette(rev(brewer.pal(n = 9, name =  "Spectral")))(length(mat_breaks_avg) - 1),
 breaks            = mat_breaks_avg,
  clustering_method = 'average',
  cluster_cols = F,
  cluster_rows = T,
  drop_levels = TRUE,
  border_color = NA,
  fontsize = 8,
  display_numbers = T,
  filename = "outputs/figures/heatmap_top50.pdf", width = 5, height = 8
  )

  mat_breaks_avg <- quantile_breaks(as.matrix(Genus_top50_AVG), n = 1000)
pheatmap(
  mat               = as.matrix(Genus_top50_AVG),
  color             = colorRampPalette(rev(brewer.pal(n = 9, name =  "Spectral")))(length(mat_breaks_avg) - 1),
 breaks            = mat_breaks_avg,
  clustering_method = 'average',
  cluster_cols = F,
  cluster_rows = T,
  drop_levels = TRUE,
  border_color = NA,
  fontsize = 8,
  display_numbers = T, number_format = "%.1e",
  filename = "outputs/figures/heatmap_top50_AVG.pdf", width = 7, height = 5
  )

```

*******************************************************************
*******************************************************************
*******************************************************************
ALPHA DIVERSITY
*******************************************************************
*******************************************************************
*******************************************************************

DIVERSITY
```{r}

# Calculate Shannon diversity
shannon_diversity <- diversity(ASV_table_data, index = "shannon") #ASV_table_data is your raw data table. 
shannon_by_group <- tapply(shannon_diversity, fact$treatment, mean)
# Fit ANOVA model
anova_model <- aov(shannon_diversity ~ fact$treatment)
# Print the ANOVA results
summary(anova_model)

# Perform Kruskal-Wallis test
kruskal_test_result <- kruskal.test(shannon_diversity ~ fact$treatment)
# Print the Kruskal-Wallis test results
kruskal_test_result
# Perform Tukey's HSD test for post-hoc comparisons
tukey_results <- TukeyHSD(anova_model)
# Print the results
print(tukey_results)
```

SPECIES RICHNESS

```{r}

# Calculate species richness
species_richness <- specnumber(ASV_table_data)
fact$richness <- species_richness
richness_by_group <- tapply(species_richness, fact$treatment, mean)
# Fit ANOVA model
anova_model_rich <- aov(species_richness ~ fact$treatment)
# Print the ANOVA results
summary(anova_model_rich)

# Perform Kruskal-Wallis test
kruskal_test_result_rich <- kruskal.test(species_richness ~ fact$treatment)
# Print the Kruskal-Wallis test results
kruskal_test_result_rich
# Perform Tukey's HSD test for post-hoc comparisons
tukey_results <- TukeyHSD(anova_model_rich)
# Print the results
print(tukey_results)
```
Plot DIVERSITY and RICHENSSS

```{r}
myPalette <- c("#C0ECB3", "#7A8985", "#D0DFE0", "#66E79E", "#DED1A0", "#71AA64" ,"#82E754", "#78E1D8" ,"#7D73D4", "#D4A6DC", "#DD9FA7", "#DC5F5B", "#83B6E1", "#E2A754" ,"#D965B2" ,"#DAE25F" ,"#BF47E2")


data_summary <- function(data, varname, groupnames){
  require(plyr)
  summary_func <- function(x, col){
    c(mean = mean(x[[col]], na.rm=TRUE),
      sd = sd(x[[col]], na.rm=TRUE))
  }
  data_sum<-ddply(data, groupnames, .fun=summary_func,
                  varname)
  data_sum <- rename(data_sum, c("mean" = varname))
  return(data_sum)
}

inv_simp <- diversity(ASV_deseq_norm, index = "invsimpson")
fact$invsimpson <- inv_simp

df_simp <- data_summary(fact, varname="invsimpson", 
                    groupnames=c("treatment"))

# Convert dose to a factor variable
df_simp$treatment <- as.factor(df_simp$treatment)
head(df_simp)

y_limit <- max(df_simp$invsimpson + df_simp$sd)

plotSimp <- ggplot(df_simp, aes(treatment, invsimpson)) +
  geom_bar(stat="identity", color="black", fill = "#78E1D8") +
  theme_bw() +
  theme(axis.text.x = element_text(size = 8)) +
  theme(plot.title = element_text(hjust = 0.5)) +
  theme(legend.position = "bottom") +
  theme(legend.title = element_blank()) +
  scale_y_continuous(expand = c(0,0), limits = c(0, y_limit + 5)) +
  xlab("") +
  ylab("Inverse Simpson") + 
  theme(panel.grid = element_blank()) +
  geom_errorbar(aes(ymin = invsimpson, ymax = invsimpson+sd), width = .2,
                position = position_dodge(.9)) 

ggsave(plot = plotSimp , filename = "outputs/figures/InvSimp_treatment.pdf", width = 5, height = 5)

```


```{r}
df2 <- data_summary(fact, varname="richness", 
                    groupnames = c("treatment"))

df_rich <- data_summary(fact, varname = "richness", 
                    groupnames = c("treatment"))
# Convert dose to a factor variable
df_rich$treatment <- as.factor(df_rich$treatment)
head(df_rich)

y_limit <- max(df2$richness + df2$sd)

plotRichn <- ggplot(df_rich, aes(treatment, richness)) +
  geom_bar(stat="identity", color="black", fill = "#78E1D8") +
  theme_bw() +
  theme(axis.text.x = element_text(size = 8)) +
  theme(plot.title = element_text(hjust = 0.5)) +
  theme(legend.position = "bottom") +
  theme(legend.title = element_blank()) +
  scale_y_continuous(expand = c(0,0), limits = c(0, y_limit + 50)) +
  xlab("") +
  ylab("Number of ASV sequences") + 
  theme(panel.grid = element_blank()) +
  geom_errorbar(aes(ymin = richness, ymax = richness+sd), width = .2,
                position = position_dodge(.9)) 
plotRichn
ggsave(plot = plotRichn , filename = "outputs/figures/Richn_treatment.pdf", width = 5, height = 5)

```

*******************************************************************
*******************************************************************
*******************************************************************
DISSIMILARITY MATRIX and NMDS BASED ON SORENSEN (presence/absence)
*******************************************************************
*******************************************************************
*******************************************************************

One method to calculate Sorensen using package betapart
First I will transform my abundance table to presence/absence
**beta.pair function**
beta.sim	
dist object, dissimilarity matrix accounting for spatial turnover (replacement), measured as Simpson pair-wise dissimilarity

beta.sne	
dist object, dissimilarity matrix accounting for nestedness-resultant dissimilarity, measured as the nestedness-fraction of Sorensen pair-wise dissimilarity

beta.sor	
dist object, dissimilarity matrix accounting for total dissimilarity, measured as Sorensen pair-wise dissimilarity (a monotonic transformation of beta diversity)
```{r}

head(dat_Abs_filtered)
dat_Abs_filtered_PA <- dat_Abs_filtered %>% mutate_if(is.numeric, ~1 * (. != 0)) #transform my abundance table to presence/absence
head(dat_Abs_filtered_PA)


library(betapart)
  beta <- beta.pair(dat_Abs_filtered_PA, index.family="sorensen")
str(beta)

m.sim <- as.matrix(beta$beta.sim)
m.sne <- as.matrix(beta$beta.sne)
m.sor <- as.matrix(beta$beta.sor)
#write.csv(m.sim, "BLCA_Default_settings_changed_outputs/outputs/files/sim.csv")
#write.csv(m.sne, "BLCA_Default_settings_changed_outputs/outputs/files/sne.csv")
#write.csv(m.sor, "BLCA_Default_settings_changed_outputs/outputs/files/sor.csv")


m.sor <-  vegdist(m.sor)
beta_mds <- metaMDS(m.sor, trymax = 1000)
beta_mds # stress = 0.06

# NMDS plot
MDS_xy_sor <- data.frame(beta_mds$points)
MDS_xy_sor$Treatment <- fact$Treatment_Time

# NMDS plot with Treatment and Time as individual factors
MDS_xy_Treat_Time_sor <- data.frame(MDS1 = beta_mds$point[,1], MDS2 = beta_mds$point[,2], Treatment_Time = as.factor(fact$Treatment_Time), Treatment = as.factor(fact$Treatment), Time = as.factor(fact$Time))

# plot and save
shape_values <- 15:17

plot_sor <- ggplot(MDS_xy_Treat_Time_sor, aes(MDS1, MDS2)) +
  geom_point(aes(color = Treatment, shape = Time), size = 3) +
  scale_color_manual(values = c("#A6DBA0","#BF812D", "#C2A5CF")) +
  scale_shape_manual(values = shape_values) +
  xlab("NMDS1") +
  ylab("NMDS2") +
  theme_bw() +
  theme(panel.grid = element_blank()) +
  theme(aspect.ratio = 1)

#ggsave(plot = plot_sor, file = "outputs/outputs/figures/NMDS_allOTUs_Treatment_Sorensen.pdf", width = 6, height = 5)

```

**STATS on dissimilarity matrix -- presence/absence - Sorensen -- **

```{r}

adonis(m.sor ~ fact$Treatment)
heatmap(as.matrix(m.sor))

rc <- rainbow(nrow(x), start = 0, end = .3)

pdf(file = "outputs/outputs/figures/heatmap_sorensen.pdf")
heatmap(as.matrix(m.sor), Rowv = NA, Colv = NA, scale = "column",
        main = "")
dev.off()


#PERMANOVA with Adonis
PW.Adonis_sorensen = pairwise.adonis(dat_Abs_filtered_PA, factors, sim.method="jaccard",p.adjust.m = "bonferroni")
#write.csv(PW.Adonis_sorensen,"BLCA_Default_settings_changed_outputs/outputs/files/Adonis-Results_Sorensen.csv")

```

Now I will use the data table of dissimilarities D2 (Treatment1) vs C and D3 (Treatment 2) vs C. I manually edited it to be able to get it. 
What I want to do is to do the stats on the dissimilarities of the two treatments vs the control. 
**SORENSEN**
```{r}
data_SORdissimilarities <- read.csv("outputs/outputs/files/sor_dissimilaritiesD2vsD3.csv", row.names = 1)
fact_dissimilarities <- read.csv("outputs/outputs/files/metadataDissimilarities_Ulva.csv", row.names = 1)

adonis(data_SORdissimilarities ~ fact_dissimilarities$Treatment * fact_dissimilarities$Time)
 
adonis(data_SORdissimilarities ~ fact_dissimilarities$Treatment_Time)

factors=fact_dissimilarities$Treatment_Time
PW.AdonisSOR_D2vsD323 = pairwise.adonis(data_SORdissimilarities,factors,sim.method="bray", p.adjust.m = "bonferroni")
#write.csv(PW.AdonisSOR_D2vsD3,"outputs/files/AdonisD2vsD3Dissimilarities-Results_Sorensen.csv")
```


BOXPLOT of dissimilarity matrix values
**Bray Curtis**
```{r}
library(wesanderson)


bc_TreatmentVSControl <-  read.csv("outputs/files/bc_TreatmentsVScontrol.csv", row.names = 1)
head(bc_TreatmentVSControl)
head(fact)

fact_treatments <-  fact[-c(1:18),]

#bc_TreatmentVSControl_final <-  merge(bc_TreatmentVSControl, fact_treatments, by = "sample")

bc_TreatmentVSControl <- t(bc_TreatmentVSControl) %>%  as.data.frame() %>% rownames_to_column(var = "sample") 

bc_long <- bc_TreatmentVSControl %>%
  tidyr::gather(key = "sample", value = "bc_values")
head(bc_long)

# merge tables
bc_final <- bc_long %>%
  dplyr::select(sample, bc_values) %>%
  left_join(., fact_treatments, by = "sample")
head(bc_final)
tail(bc_final)


bc_plot <- bc_final %>%
  dplyr::group_by(sample, Treatment) %>%
  dplyr::summarise(sum_bc = sum(bc_values))

bc_avg <- data.frame(bc_final %>%
                                  dplyr::group_by(sample, Treatment) %>%
                                  dplyr::summarise(mean_bc = mean(bc_values)))

 boxplot_bc <- ggplot(bc_avg, aes(Treatment, mean_bc, fill = Treatment)) +
  geom_boxplot() +
  theme_bw() +
  theme(plot.title = element_text(size = 6.5, hjust = 0.5)) +
  theme(legend.position = "none") + 
  #theme(legend.title = element_blank()) + 
  #theme(legend.title = element_text(size = 5)) +
  #scale_y_continuous(expand = c(0,0), labels = scales::percent) +
  #scale_fill_manual("legend", values = myPalette_replicates) +
  #scale_fill_manual(values=wes_palette(n=3, name="Royal1")) + 
  scale_fill_grey() +
  xlab("") +
  ylab("Bray-Curtis Dissimilarity") + 
  facet_grid(.~ Treatment, scales = "free") +
  theme(axis.text.x = element_text(size = 6.5, angle = 0, hjust = 0, vjust = 0)) +
   theme(axis.text.y = element_text(size = 6.5, angle = 0, hjust = 0, vjust = 0)) +
  theme(panel.grid = element_line()) 

#ggsave(plot = boxplot_bc, file = "outputs/figures/Boxplot_Bray_Curtis.pdf", width = 7, height = 5)
 #ggsave(plot = boxplot_bc, file = "outputs/figures/Boxplot_Bray_Curtis_GRAYSCALE_ISME.pdf",   width = 80,  ###just make sure <175
      # height =70,  ###can change depend on how your plot looks like
      # units = "mm")
```

**Sorensen**
```{r}
sor_TreatmentVSControl <-  read.csv("outputs/files/sor_TreatmentsVScontrol.csv", row.names = 1)
head(bc_TreatmentVSControl)
head(fact)

fact_treatments <-  fact[-c(1:18),]

#sor_TreatmentVSControl_final <-  merge(bc_TreatmentVSControl, fact_treatments, by = "sample")

sor_TreatmentVSControl <- t(sor_TreatmentVSControl) %>%  as.data.frame() %>% rownames_to_column(var = "sample") 

sor_long <- sor_TreatmentVSControl %>%
  tidyr::gather(key = "sample", value = "sor_values")
head(sor_long)

# merge tables
sor_final <- sor_long %>%
  dplyr::select(sample, sor_values) %>%
  left_join(., fact_treatments, by = "sample")
head(sor_final)
tail(sor_final)


sor_plot <- sor_final %>%
  dplyr::group_by(sample, Treatment) %>%
  dplyr::summarise(sum_sor = sum(sor_values))

sor_avg <- data.frame(sor_final %>%
                                  dplyr::group_by(sample, Treatment) %>%
                                  dplyr::summarise(mean_sor = mean(sor_values)))

 boxplot_sor <- ggplot(sor_avg, aes(Treatment, mean_sor, fill = Treatment)) +
  geom_boxplot() +
  geom_boxplot() +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5)) +
  theme(legend.position = "none") + 
  #theme(legend.title = element_blank()) + 
  #theme(legend.title = element_text(size = 8)) +
  #scale_y_continuous(expand = c(0,0), labels = scales::percent) +
  #scale_fill_manual("legend", values = myPalette_replicates) +
  #scale_fill_manual(values=wes_palette(n=3, name="Royal1")) + 
  scale_fill_grey() +
  xlab("") +
  ylab("Sorensen Dissimilarity") + 
  facet_grid(.~ Treatment, scales = "free") +
  theme(axis.text.x = element_text(size = 6.5, angle = 0, hjust = 0, vjust = 0)) +
   theme(axis.text.y = element_text(size = 6.5, angle = 0, hjust = 0, vjust = 0)) +
  theme(panel.grid = element_line()) 

#ggsave(plot = boxplot_sor, file = "outputs/figures/Boxplot_Sorensen.pdf", width = 7, height = 5)
 #ggsave(plot = boxplot_sor, file = "outputs/figures/Boxplot_Sorensen_GRAYSCALE_ISME.pdf", width = 80,  ###just make sure <175
      # height =70,  ###can change depend on how your plot looks like
     #  units = "mm")
```

